aten/src/ATen/native/cuda/NativeFunctionsCuda.cu

```diff
@@ -0,0 +1,210 @@
+    int pw = linearIndex % pooledWidth;
+    int ph = (linearIndex / pooledWidth) % pooledHeight;
+    int ch = (linearIndex / pooledWidth / pooledHeight) % inputChannels;
+    int proposal = linearIndex / pooledWidth / pooledHeight / inputChannels;
+
+    // Get particular proposal data
+    const T *roisOffset = rois + (proposal * 5);
+    int n = roisOffset[0];
+    int startWidth = llrintf(roisOffset[1] * spatialScale);
+    int startHeight = llrintf(roisOffset[2] * spatialScale);
+    int endWidth = llrintf(roisOffset[3] * spatialScale);
+    int endHeight = llrintf(roisOffset[4] * spatialScale);
+
+    // TODO: fix malformed RoIs to be 1x1
+
+    int roiHeight = endHeight - startHeight;
+    int roiWidth = endWidth - startWidth;
+
+    // Calculate size of tile based on the size of this particular RoI and the
+    // output size
+    T tileHeight = static_cast<T>(roiHeight) / static_cast<T>(pooledHeight);
+    T tileWidth = static_cast<T>(roiWidth) / static_cast<T>(pooledWidth);
+
+    // Calculate offset into the pooled region
+    int tileHStart = static_cast<int>(floorf(static_cast<T>(ph) * tileHeight));
+    int tileWStart = static_cast<int>(floorf(static_cast<T>(pw) * tileWidth));
+    int tileHEnd = static_cast<int>(ceilf(static_cast<T>(ph + 1) * tileHeight));
+    int tileWEnd = static_cast<int>(ceilf(static_cast<T>(pw + 1) * tileWidth));
+
+    // Calculate offset into the image itself, based on RoI + pooled offsets,
+    // and ensure it falls within image boundaries
+    tileHStart = fmin(fmax(tileHStart + startHeight, 0), inputHeight);
+    tileWStart = fmin(fmax(tileWStart + startWidth, 0), inputWidth);
+    tileHEnd = fmin(fmax(tileHEnd + startHeight, 0), inputHeight);
+    tileWEnd = fmin(fmax(tileWEnd + startWidth, 0), inputWidth);
+
+    // If our pooling region is empty, we set the output to 0, otherwise to
+    // the min float so we can calculate the max properly
+    bool isEmpty = (tileHStart >= tileHEnd) || (tileWStart >= tileWEnd);
+    T max = isEmpty ? 0 : FLT_MIN;
+    // If nothing is pooled, argmax = -1 causes nothing to be backprop'd
+    int maxIdx = -1;
+
+    const T *inputOffset = input + ((n * inputChannels + ch) * inputHeight * inputWidth);
+    for (int th = tileHStart; th < tileHEnd; ++th) {
+      for (int tw = tileWStart; tw < tileWEnd; ++tw) {
+        int index = (th * inputWidth) + tw;
+	if (inputOffset[index] > max) {
+          max = inputOffset[index];
+	  maxIdx = index;
+	}
+      }
+    }
+    output[linearIndex] = max;
+
+    // TODO optional argmax
+    argmaxes[linearIndex] = maxIdx;
+  }
+}
+
+std::tuple<Tensor, Tensor> RoiPooling2d_forward_cuda(
+  const Tensor& input,
+  const Tensor& rois,
+  int64_t pooledHeight,
+  int64_t pooledWidth,
+  double spatialScale)
+{
+
+  // Input is the output of the last convolutional layer in the Backbone network, so
+  // it should be in the format of NCHW
+  AT_ASSERT(input.ndimension() == 4, "Input to RoI Pooling should be a NCHW Tensor");
```

Comment:
@ezyang has some good helpers in his cuDNN work that will improve these error messages. No need for anything this PR but it is something to be aware of when it lands.

Reply:
Yep. These kernels aren't really optimized or clean in anyway, mainly just copied from Caffe2. They will be improved upon later I just need something that works.